Fix unlock the pruned partitions of partition table

pobbatihari opened the pull request:
FIXME's:

     * GPDB_12_MERGE_FIXME: Revisit this post merge and see if we have a
       chance to unlock the pruned partitions
     
     * GPDB_12_MERGE_FIXME: Revisit this post merge and see if we have a
       chance to unlock the pruned partitions

RCA:
	ORCA always require root partition statistics for accurate query
	planning and cost estimation. While fetching total size of a
	partition table we lock all the leaf partitions including pruned
	parittions, but if we find the relation statistics in MD cache,
	we only acquire lock on root partition table. Although acquire
	lock on every leaf partition is unnecessary, we have to acquire
	lock for necessary leaf partitions on master node, otherwise it
	can lead to deadlock issues.

	For example :
	//Assuming relation stats are available at MD Cache, so ORCA
	//acquires only lock on root parittion.
	txn A: select * from root_part;
	txn B: truncate leaf1;
	txn A and B will not block each other on QD if no locking leaf
	suppose on seg0 txn A arrive first, seg1 txn B arrive first
	on seg0, txn B will wait for A
	on seg1, txn A will wait for B
	Global Deadlock happens

Fix:
	This PR fixes the unnecessary lock on leaf partitions of a
	partition table and acquires the lock only on unpruned leaf
	partition.
	`Thumb rule is if we are going to hold a lock of table on
	segment, we should first hold the lock on master (QD)`.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

THANATOSLAVA commented:
> txn A and B will not block each other on QD if no locking leaf

I am not following why the two transactions won't block each other on the master.

If A acquires the access share lock, it should block truncate.

If B acquires the access exclusive lock, it should block all other transactions accessing the partition table.

kainwen commented:
> > txn A and B will not block each other on QD if no locking leaf
> 
> I am not following why the two transactions won't block each other on the master.
> 
> If A acquires the access share lock, it should block truncate.
> 
> If B acquires the access exclusive lock, it should block all other transactions accessing the partition table.

Hi @THANATOSLAVA you can refer to this old issue  https://github.com/greenplum-db/gpdb/issues/13652
the bug pattern is: a session hold locks on segment without the same lock on coordinator. The truncate is directly on a leaf partition however the select is on root.

kainwen commented on src/test/regress/expected/partition_locking_optimizer.out:
```diff
@@ -263,11 +286,12 @@ insert into partlockt values(1), (2), (3);
 select * from locktest_master where coalesce not like 'gp_%' and coalesce not like 'pg_%';
      coalesce      |       mode       | locktype |  node  
 -------------------+------------------+----------+--------
+ partlockt         | AccessShareLock  | relation | master
```
Why one more `AccessShareLock`?

pobbatihari replied:
In ORCA we are seeing discrepancy with 'AccessShareLock' on root partition table
with-assert and without-assert modes of DML queries, because
with-assert mode we are forced to release the cache, which leads to acquire 'AccessShareLock' always on master.

partcache.c : generate_partition_qual():335

```
	/* If we already cached the result, just return a copy */
	if (rel->rd_partcheckvalid)
		return copyObject(rel->rd_partcheck);

	/* Grab at least an AccessShareLock on the parent table */
	parent = relation_open(get_partition_parent(RelationGetRelid(rel)),
						   AccessShareLock);
```

THANATOSLAVA commented:
> #13652

Thanks for the background. So on the master, `select root` doesn't acquire locks on the leaves. But shouldn't `truncate leaf` acquire the lock on the leaves and the root? So if B acquires the lock, it should block A on the master right?

pobbatihari commented:
> > #13652
> 
> Thanks for the background. So on the master, `select root` doesn't acquire locks on the leaves. But shouldn't `truncate leaf` acquire the lock on the leaves and the root? So if B acquires the lock, it should block A on the master right?
On master:
       `select root` doesn't acquire locks on the leaves.
         -- yes
        shouldn't `truncate leaf` acquire the lock on the leaves and the root
        -- Since `select root` not acquired any locks on leaves on master, trucate leaf can acquire lock on leaf
thats why txn A and B will not block each other on QD if no locking leaf

pobbatihari commented:
NOTE:Modified FIXME description
--- GPDB_12_MERGE_FIXME Revisit this post merge and see if we have a chance to unlock the root table
to
+-- GPDB_12_MERGE_FIXME Revisit this post merge and see why the leaf acquires AccessShareLock instead ExclusiveLock

As discussed its ok to have lock on root table, and even planner does the same thing. One thing observed is leaf partition is acquiring AccessShareLock instead Exclusive lock, updated the FIXME message according to that. 
```
ORCA:
# select * from locktest_segments where coalesce not like ‘gp_%’ and coalesce not like ‘pg_%‘;
     coalesce      |      mode       | locktype |   node
-------------------+-----------------+----------+-----------
 partlockt         | ExclusiveLock   | relation | 1 segment
 partlockt_1_prt_4 | AccessShareLock | relation | 1 segment
postgres:
# select * from locktest_segments where coalesce not like ‘gp_%’ and coalesce not like ‘pg_%‘;
     coalesce      |      mode       | locktype |   node
-------------------+-----------------+----------+-----------
 partlockt         | AccessShareLock | relation | 1 segment
 partlockt         | ExclusiveLock   | relation | 1 segment
 partlockt_1_prt_4 | ExclusiveLock   | relation | 1 segment
```

THANATOSLAVA commented:
> shouldn't `truncate leaf` acquire the lock on the leaves and the root
> -- Since `select root` not acquired any locks on leaves on master, trucate leaf can acquire lock on leaf
> thats why txn A and B will not block each other on QD if no locking leaf

Ah what I meant was, on the master, shouldn't `truncate leaf` attempt to acquire the lock on both leaves and the root. If it manages to acquire lock on both, it should block `select root`, cause the latter would also attempt to acquire lock on the root.

THANATOSLAVA commented:
> Fix:
This PR fixes the unnecessary lock on leaf partitions of a
partition table and acquires the lock only on unpruned leaf
partition.

It makes sense to acquire locks on all the partitions to be scanned when building the dynamic scan nodes, so we block leaf modification attempts by other processes.

But while retrieving the stats from relcache, shouldn't we still be acquiring access share locks on all the partitions? Why is this no more necessary?

pobbatihari commented:
> > Fix:
> > This PR fixes the unnecessary lock on leaf partitions of a
> > partition table and acquires the lock only on unpruned leaf
> > partition.
> 
> It makes sense to acquire locks on all the partitions to be scanned when building the dynamic scan nodes, so we block leaf modification attempts by other processes.
> 
> But while retrieving the stats from relcache, shouldn't we still be acquiring access share locks on all the partitions? Why is this no more necessary?

We actually acquire lock for retrieving  reltuples for the leaf partitions, by this time we will be holding lock on root partition. So we thought AccessShareLock might not be necessary and the lock on root partition will block concurrent transitions like alter, drop, vacuum, etc . But yes still truncate on leaf can create some mess on calculating total size of table.  
I think  acquire lock while retrieving reltuples and release the lock by end of the block (cdb_estimate_partitioned_numtuples) might be better. will work on it.

chrishajas commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -706,7 +706,7 @@ cdb_estimate_partitioned_numtuples(Relation rel)
 
 		if (childrel != rel)
-			heap_close(childrel, NoLock);
+			heap_close(childrel, AccessShareLock);
```
In L679, we open this relation with NoLock. Why are we closing it with an AccessShareLock? This needs comments. Please also add a comment on why we need an access share lock of the leaf partitions

pobbatihari replied:
Added more comments

THANATOSLAVA replied:
So we acquire access share locks on all the partitions while retrieving stats from relcache. Just to be sure, if we don't release the locks here, they would be held until the end of transaction?

pobbatihari replied:
yes, you are correct

THANATOSLAVA replied:
I wonder if there's value in holding the locks on all the leaves until we build the table scan plan nodes. By then we know which leaves to scan and which ones to prune. And we can release the locks on all the leaves, and just acquire the locks on the leaves to be scanned. I guess the question is, between exploration/implementation and planned statement, do we need to worry about the stats being updated by another query.

pobbatihari replied:
Yes, During the implementation/optimization phase, the totaltuples in the leaf partitions is accessible to other transactions, and we acquire locks on unpruned partitions during plan statements.

The rationale for releasing these locks is as follows:
- AccessShareLock on all leaf partitions is unnecessary during DML/DDL operations.
- Keeping these locks unnecessarily blocks the load on the leaf partitions.

THANATOSLAVA replied:
Scrape off everything that I said above. Had a discussion with the team, and realized we definitely prioritize concurrency over up-to-date stats, i.e. better plans. So like @dgkimura suggested, it's okay that we don't acquire locks at all when we are retrieving stats from rel cache.

chrishajas replied:
I agree with David too, I don't think we need to take a lock here

pobbatihari replied:
Thanks for the update @THANATOSLAVA , @chrishajas

pobbatihari replied:
created a PR (#15485 )for removing asserts in release build at CTranslatorDXLToPlStmt.cpp

THANATOSLAVA commented:
> on the master, `select root` attempts to acquire the lock on the root, and `truncate leaf` attempts to acquire the lock on both the leaves and the root. 
> Now, if `truncate leaf` manages to acquire the locks on both, that should block `select root`, as the lock on the root is currently owned by `truncate leaf`. In that sense, `truncate leaf` should be able to block `select root` on the master, right?

Here's to summarize the discussion with @pobbatihari. The following all refers to the behavior on the master.

1. Because the stats are available on the metadata cache, `select root` only acquires access share lock on the root, and not any locks on the leaves.
2. `truncate leaf` acquires access share lock on the root, and access exclusive lock on the particular leaf, and not any locks on the other leaves.

Both transactions acquire shared lock on the root. That's why `select root` and `truncate leaf` won't block each other on the master.

dgkimura commented on src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp:
```diff
@@ -4090,10 +4090,17 @@ CTranslatorDXLToPlStmt::TranslateDXLDynTblScan(
 
+	const CDXLTableDescr *dxl_table_descr =
+		dyn_tbl_scan_dxlop->GetDXLTableDescr();
+	GPOS_RTL_ASSERT(dxl_table_descr->LockMode() != -1);
```
AFAIK `GPOS_RTL_ASSERT` is enabled in all builds (release and debug). Why do we want/need this assert to be enabled in release builds?

pobbatihari replied:
Yes, you are correct GPOS_RTL_ASSERT is enabled in all builds. However, I am unsure if assertion can only be used in debug builds. I have followed the approach adopted by other operators (Translate DXLTblScan). Let me know ur thoughts

dgkimura replied:
I think we need to figure out how the assert could ever be false.  And if the assert is false, what's the worst case scenario (e.g. crash or wrong results). We probably should also do some archeology to try to understand why the other places used `GPOS_RTL_ASSERT` instead of just `GPOS_ASSERT`.

The way I see it, assertions are developer tools to validate assumptions made by the code.  And if it's not true, then the intent is to let the developer know.  If that's the only scenario, then I don't see why retail builds should enable asserts.

pobbatihari replied:
The rte->rellockmode value of -1 in PostgreSQL indicates that the lock mode for a relation is not explicitly specified or is unknown. 
lockmode()(rte->rellockmode) which is set by the parser based on various factors  such as the type of operation being 
performed on the relation, the transaction isolation level, and concurrency control considerations, etc..

In this scenario, we are checking the lock mode of a table that is going to be scanned. In my understanding, it is appropriate to have the assertion/error 
even in the release build. If we don't assert/error in the release build, the plan may be executed without the proper lock on the table in QD and with locks in QE, 
which can lead to potential issues with concurrent transactions. 

comments from code:
// Lock any table we are to scan, since it may not have been properly locked
// by the parser (e.g in case of generated scans for partitioned tables)

dgkimura replied:
IIUC, you're arguing that it's normal that the value may be -1 at this point in the code.  And you're using the assert to bail out, which is essentially a fall back to PLANNER.  If that's the case, then why not use GPOS_RAISE with an error message explaining that?  But even if we did that, who's to say that PLANNER doesn't have the same issue?  ISTM, that maybe what you really want is `elog(ERROR, ...)` from ORCA.  Also, are is there a test covering this scenario?

chrishajas replied:
In `LockAcquireExtended`, there's 
```
	if (lockmode <= 0 || lockmode > lockMethodTable->numLockModes)
		elog(ERROR, "unrecognized lock mode: %d", lockmode);
```

So -1 seems to be an invalid value. If this is an assumption that we're making, I think it should be a regular (non retail) assert. Having asserts in retail builds ends up slowing things down. If it's a valid, we should have an explicit fallback. If it's being ultra-defensive, we should have an explicit fallback using GPOS_RAISE, but that also needs an explanation of how/why we might get into this scenario.

gpopt replied:
@pobbatihari I agree with @chrishajas and @dgkimura.

pobbatihari replied:
Understood. Thank you, David and Chris, for your input. I will update as suggested

dgkimura replied:
I think it makes sense to create a new pull-request for higher visibility.  At the very least, it should probably be merged as a separate commit.  I also noticed that Alex made one of those changes, so it would probably be good to get buy-in from her as well.

pobbatihari replied:
sure @dgkimura . I will create a new PR for to accommodate these change for other operators.

pobbatihari replied:
created a PR (#15485 )for removing asserts in release build at CTranslatorDXLToPlStmt.cpp

dgkimura commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -665,6 +665,12 @@ cdb_estimate_partitioned_numtuples(Relation rel)
 	if (rel->rd_rel->reltuples > 0)
 		return rel->rd_rel->reltuples;
 
+	// To prevent any concurrent transaction from causing discrepancies in
+	// the computation of the relation totaltuples, it is recommended to
+	// obtain an AccessShareLock on the leaf partitions. This lock will
+	// ensure that any truncation operation performed by concurrent
+	// transactions on the leaf partitions does not interfere with the
+	// accuracy of the totaltuples calculation.
 	inheritors = find_all_inheritors(RelationGetRelid(rel),
```
Curious, is this really a big deal?  I understand that without locking we may use bad stats to pick a bad plan, but how often does it happen?  And if we take the lock, how much might it slow down other transactions in the worst case scenario?  And if it's really a big deal, then why can't the user specify a different transaction isolation level (e.g. "serializable" instead of "read committed") in the cases where it matters?

pobbatihari replied:
The introduction of this lock occurred in version 7X, and it occurs once during every query when deriving statistics at the FinalizeExploration() stage. I am not sure how much slows dow other transactions in the worst-case scenario.  AFAIK, it blocks concurrent transactions (truncate, delete, etc) on the leaf partitions until the transaction is completed.

THANATOSLAVA commented:
> This PR fixes the unnecessary lock on leaf partitions of a
partition table and acquires the lock only on unpruned leaf
partition.

Since this PR targets two issues, racing condition (global dead lock) and inefficiency (locking pruned partitions), it would be helpful to address the two issues separately in the description of the fix.

Here's my understanding --

1. To avoid the global dead lock, we acquire the locks on the partitions to be scanned when ORCA writes the plan, regardless of the stats being in metadata cache or not. This way, if another transaction attempts to edit a leaf partition to be scanned by the current session, the other transaction would be blocked on the master.
2. To resolve the inefficiency in locking pruned partitions, we decide to not acquire locks at all while retrieving stats from the rel cache. Worst case, we form a plan based on obsolete stats by the time of execution, and that's still better than sacrificing concurrency. In the meantime, we only acquire the locks on the partitions to be scanned when ORCA writes the plan. This way, if another transaction attempts to edit a leaf partition pruned by the current session, the other transaction would be able to proceed.

pobbatihari commented:
> > This PR fixes the unnecessary lock on leaf partitions of a
> > partition table and acquires the lock only on unpruned leaf
> > partition.
> 
> Since this PR targets two issues, racing condition (global dead lock) and inefficiency (locking pruned partitions), it would be helpful to address the two issues separately in the description of the fix.
> 
> Here's my understanding --
> 
> 1. To avoid the global dead lock, we acquire the locks on the partitions to be scanned when ORCA writes the plan, regardless of the stats being in metadata cache or not. This way, if another transaction attempts to edit a leaf partition to be scanned by the current session, the other transaction would be blocked on the master.
> 2. To resolve the inefficiency in locking pruned partitions, we decide to not acquire locks at all while retrieving stats from the rel cache. Worst case, we form a plan based on obsolete stats by the time of execution, and that's still better than sacrificing concurrency. In the meantime, we only acquire the locks on the partitions to be scanned when ORCA writes the plan. This way, if another transaction attempts to edit a leaf partition pruned by the current session, the other transaction would be able to proceed.

Thank you @THANATOSLAVA  for the detailed summary.   I acknowledge that the pull request addresses two separate issues. However, I noticed that if we handle these fixes separately that we end up editing the output of the same test cases, which seems redundant and adds unnecessary extra work. Considering the minimal code changes 
involved, would it be possible to proceed with a single pull request for both issues?

gpopt commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -676,7 +678,7 @@ cdb_estimate_partitioned_numtuples(Relation rel)
 
 		if (childid != RelationGetRelid(rel))
-			childrel = try_table_open(childid, NoLock, false);
+			childrel = childrel = RelationIdGetRelation(childid);
```
`childrel = childrel = `?

pobbatihari replied:
updated.

gpopt reviewed (approved):
+1

chrishajas commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -676,7 +678,7 @@ cdb_estimate_partitioned_numtuples(Relation rel)
+			childrel = RelationIdGetRelation(childid);
 		else
 			childrel = rel;
 
```
What happens if `childrel` is NULL? I think the existing code had that issue, but I think in that case we want to just assume 0 rows for the child and continue the loop

dgkimura commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -665,9 +665,11 @@ cdb_estimate_partitioned_numtuples(Relation rel)
-	inheritors = find_all_inheritors(RelationGetRelid(rel),
-									 AccessShareLock,
-									 NULL);
+	// To avoid blocking concurrent transactions on leaf partitions
+	// throughout the entire transition, we refrain from acquiring locks on
+	// the leaf partitions. Instead, we acquire locks only on the
+	// partitions that need to be scanned when ORCA writes the plan.
```
I think it's worth explicitly calling out that this could lead to less accurate stats.  But that's a trade-off we knowingly accept.

dgkimura reviewed (approved):
Looks good to me!

gpopt commented on src/backend/optimizer/util/plancat.c:
```diff
@@ -676,10 +678,15 @@ cdb_estimate_partitioned_numtuples(Relation rel)
 
+		// If childrel is NULL, continue by assuming the child relation
+		// has 0 tuples.
+		if (NULL == childrel) {
+			continue;
+		}
```
Oh, @pobbatihari , in c files, we use `childrel == NULL` instead of `NULL == childrel`.
